virtcontainers/device/drivers/generic.go

```diff
@@ -27,38 +32,98 @@ func NewGenericDevice(devInfo *config.DeviceInfo) *GenericDevice {
 
 // Attach is standard interface of api.Device
 func (device *GenericDevice) Attach(devReceiver api.DeviceReceiver) error {
-	if device.DeviceInfo.Hotplugged {
+	skip, err := device.bumpAttachCount(true)
+	if err != nil {
+		return err
+	}
+	if skip {
 		return nil
 	}
-
+	device.AttachCount = 1
 	return nil
 }
 
 // Detach is standard interface of api.Device
 func (device *GenericDevice) Detach(devReceiver api.DeviceReceiver) error {
-	if !device.DeviceInfo.Hotplugged {
+	skip, err := device.bumpAttachCount(false)
+	if err != nil {
+		return err
+	}
+	if skip {
 		return nil
 	}
-
+	device.AttachCount = 0
 	return nil
 }
 
-// IsAttached checks if the device is attached
-func (device *GenericDevice) IsAttached() bool {
-	return device.DeviceInfo.Hotplugged
+// DeviceType is standard interface of api.Device, it returns device type
+func (device *GenericDevice) DeviceType() config.DeviceType {
+	return config.DeviceGeneric
+}
+
+// GetDeviceInfo returns device information used for creating
+func (device *GenericDevice) GetDeviceInfo() interface{} {
+	return device.DeviceInfo
+}
+
+// GetAttachCount returns how many times the device has been attached
+func (device *GenericDevice) GetAttachCount() uint {
+	return device.AttachCount
 }
 
 // DeviceID returns device ID
 func (device *GenericDevice) DeviceID() string {
 	return device.ID
 }
 
-// DeviceType is standard interface of api.Device, it returns device type
-func (device *GenericDevice) DeviceType() config.DeviceType {
-	return config.DeviceGeneric
+// GetMajorMinor returns device major and minor numbers
+func (device *GenericDevice) GetMajorMinor() (int64, int64) {
+	return device.DeviceInfo.Major, device.DeviceInfo.Minor
 }
 
-// GetDeviceInfo returns device information used for creating
-func (device *GenericDevice) GetDeviceInfo() interface{} {
-	return device.DeviceInfo
+// Reference adds one reference to device
+func (device *GenericDevice) Reference() uint {
+	if device.RefCount != intMax {
+		device.RefCount++
+	}
+	return device.RefCount
+}
+
+// Dereference remove one reference from device
+func (device *GenericDevice) Dereference() uint {
+	if device.RefCount != 0 {
+		device.RefCount--
+	}
+	return device.RefCount
+}
+
+// bumpAttachCount is used to add/minus attach count for a device
+// * attach bool: true means attach, false means detach
+// return values:
+// * skip bool: no need to do real attach/detach, skip following actions.
+// * err error: error while do attach count bump
+func (device *GenericDevice) bumpAttachCount(attach bool) (skip bool, err error) {
```

Comment:
Could you please add a lock here (to protect AttachCount calculation) so that we allow parallel hotplug of these devices?

Reply:
There's a `deviceManager` in "virtcontainers/device/manager/manager.go", it is designed to do Attach/Detach via `deviceManager.AttachDevice/DetachDevice` which has a global lock.

Do you think we need a per-device lock here? If everyone use the device manager as global entry, it should be fine to do parallel operations.

Reply:
I wonder if we could use https://golang.org/pkg/sync/atomic instead of locks?

Reply:
@jodh-intel This is a switch case with 3 conditions, I don't think we can protect it with a atomic lock.

For example, if I modify this to:

```
if attach {
    if swapped:=atomic.CompareAndSwapUint64(&device.AttachCount, 0, 0); swapped {
        return false, nil
   }else if if swapped:=atomic.CompareAndSwapUint64(&device.AttachCount, intMax, intMax); swapped {
        return true, fmt.Errorf("device was attached too many times")
    } else {
        atomic.AddUint64(&device.AttachCount, 1)
        return true, nil
    }
} else { // detach use case
...
}
```

But this logic can't protect `device.AttachCount` because between each `if...else...`, it could changed. So I don't have a good way with `Atomic`

Reply:
Yep - thanks for checking though @WeiZhang555!.